### PR TITLE
Add a Stdlib module for OCaml < 4.07

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/dune [hunk 1]
```diff
@@ -1,16 +1,72 @@
...
+let modules_post_4_02 =
+  [ "Uchar", (4, 03)
+  ; "Stdlib", (4, 07)
+  ]
+
+let missing_modules =
+  List.filter modules_post_4_02 ~f:(fun (_, v) -> version < v)
+  |> List.map ~f:fst
+
+let all_modules_expect_stdlib =
```
Comment (dbuenzli) [inline src/dune]: This should be `except` not `expect`.

**Review by dbuenzli — CHANGES_REQUESTED** [review]
I wouldn't have used `Stdlabels` I think we should rather deprecate it in the future but I won't fight against this. 

Except for the variable name this is good to go.

File: src/dune (hunk 1 above)
Comment (ghost) [reply src/dune]: Indeed, fixed

**Review by dbuenzli — APPROVED** [review]

**Comment by ghost** [comment]
> I wouldn't have used Stdlabels I think we should rather deprecate it in the future but I won't fight against this.

Out of curiosity, do you think it's the separate Stdlabels module that's bad or the use of labels in general? I don't like having to use a separate module either, but in general I think labels improve readability of the code.

**Comment by dbuenzli** [comment]
> Out of curiosity, do you think it's the separate Stdlabels module that's bad or the use of labels in general? I don't like having to use a separate module either, but in general I think labels improve readability of the code.

Both. The dichotomy present in the `stdlib` about label vs. non label makes it a pain to evolve the stdlib and one more stdlib choice to make for the programmer... Besides I dislike pointless labelling like `f` in `map` functions, if your variable names are clear they rather add reading noise in my opinion.

**Comment by ghost** [comment]
The main advantage I see of having the `f` label is that it allows re-ordering the arguments. This is especially nice when the function is big, or when you apply several transformations in a row with `|>`.

**Comment by dbuenzli** [comment]
I highly dislike the inline function style. Unless the mapped function fits on the line the function should be properly named to indicate to the reader its denotation so that the intent of the `map` (or `fold` or whatever) becomes instantaneously readable. As for chaining `|>` it's not a style that I particularly cherish but in the particular case of the classical `map`, the signature is in the right order for that.

**Comment by ghost** [comment]
I see. I understand your position but I personally don't find it unpleasant to read code written in this style, i.e. with inline functions or using `|>`.

For the case of the stdlib, given that the compiler doesn't force you to use labels and that their use is often down to personal taste, it seems to me that providing them by default would be better choice than not having them at all.

**Comment by dbuenzli** [comment]
> it seems to me that providing them by default would be better choice than not having them at all.

I disagree, the problem with labels is that they leak when you partially apply functions so they somehow go against the idea to use functions as a strong compositional glue. This makes all sorts of functional programming techniques more harder to use (e.g. lifting functions in your applicative functors).

**Comment by ghost** [comment]
> I disagree, the problem with labels is that they leak when you partially apply functions they somehow go against the idea to use functions as a strong compositional glue. This makes all sorts of functional programming techniques more harder to use (e.g. lifting functions in your applicative functors).

I understand the part about partially applying functions, and I agree that this is indeed sometimes annoying, however I don't follow the rest, I'd need to see examples to understand the problem better. 

Note that we use this style extensively at Jane Street and it works just fine for a whole lot of code.

**Pull request merged**